pkg/authorization/api/types.go

```diff
@@ -0,0 +1,110 @@
+
+const (
+	// Policy is a singleton and this is its name
+	PolicyName = "Policy"
+)
+
+// PolicyRule holds information that describes a policy rule, but does not contain information
+// about who the rule applies to or which namespace the rule applies to.
+type PolicyRule struct {
+	// Deny is true if any request matching this rule should be denied.  If false, any request matching this rule is allowed.
+	Deny bool `json:"deny" yaml:"deny"`
+	// Verbs is a list of Verbs that apply to ALL the ResourceKinds and AttributeRestrictions contained in this rule.  "*" represents all kinds.
+	Verbs []string `json:"verbs" yaml:"verbs"`
+	// AttributeRestrictions will vary depending on what the Authorizer/AuthorizationAttributeBuilder pair supports.
+	// If the Authorizer does not recognize how to handle the AttributeRestrictions, the Authorizer should report an error.
+	AttributeRestrictions kruntime.EmbeddedObject `json:"attributeRestrictions" yaml:"attributeRestrictions"`
+	// ResourceKinds is a list of kinds this rule applies to.  "*" represents all kinds.
+	ResourceKinds []string `json:"resourceKinds" yaml:"resourceKinds"`
+}
+
+// Role is a logical grouping of PolicyRules that can be referenced as a unit by RoleBindings.
+type Role struct {
+	kapi.TypeMeta   `json:",inline" yaml:",inline"`
+	kapi.ObjectMeta `json:"metadata,omitempty" yaml:"metadata,omitempty"`
+
+	Rules []PolicyRule `json:"rules" yaml:"rules"`
+}
+
+// RoleBinding references a Role, but not contain it.  It adds who and namespace information.
+// It can reference any Role in the same namespace or in the global namespace.
+type RoleBinding struct {
+	kapi.TypeMeta   `json:",inline" yaml:",inline"`
+	kapi.ObjectMeta `json:"metadata,omitempty" yaml:"metadata,omitempty"`
+
+	UserNames  []string `json:"userNames" yaml:"userNames"`
+	GroupNames []string `json:"groupNames" yaml:"groupNames"`
+
+	// Since Policy is a singleton, this is sufficient knowledge to locate a role
+	// RoleRefs can only reference the current namespace and the global namespace
+	// If the RoleRef cannot be resolved, the Authorizer must return an error.
+	RoleRef kapi.ObjectReference `json:"roleRef" yaml:"roleRef"`
+}
+
+// Policy is a object that holds all the Roles and RoleBindings for a particular namespace.  There is at most
+// one Policy document per namespace.
+// It is an attempt to mitigate referential integrity problems, but it is worth noting that it does not solve them.
+// Referential integrity problems still exist for references to the global namespace.  If that problem is resolved
+// then it seems more reasonable to destroy this resource and use the referential integrity solution to solve
+// all cases.
+type Policy struct {
+	kapi.TypeMeta   `json:",inline" yaml:",inline"`
+	kapi.ObjectMeta `json:"metadata,omitempty" yaml:"metadata,omitempty"`
+
+	LastModified kutil.Time `json:"lastModified" yaml:"lastModified"`
```

Comment:
Is there a compelling reason to add LastModified as a special case here, or should we try to get it into ObjectMeta?